src/otx/algo/detection/heads/dfine_decoder.py

```diff
@@ -0,0 +1,1165 @@
+from functools import partial
+from typing import Any, Callable, ClassVar
+
+import torch
+import torch.nn.functional as f
+from torch import Tensor, nn
+from torch.nn import init
+from torchvision.ops import box_convert
+
+from otx.algo.common.layers.transformer_layers import MLP
+from otx.algo.common.utils.utils import inverse_sigmoid
+from otx.algo.detection.heads.rtdetr_decoder import get_contrastive_denoising_training_group
+from otx.algo.utils.weight_init import bias_init_with_prob
+
+
+def weighting_function(reg_max: int, up: Tensor, reg_scale: Tensor) -> Tensor:
+    """Generates the non-uniform Weighting Function W(n) for bounding box regression.
+
+    Args:
+        reg_max (int): Max number of the discrete bins.
+        up (Tensor): Controls upper bounds of the sequence, where maximum offset is ±up * H / W.
+        reg_scale (Tensor): Controls the curvature of the Weighting Function.
+                        Larger values result in flatter weights near the central axis W(reg_max/2)=0
+                        and steeper weights at both ends.
+        deploy (bool): If True, uses deployment mode settings.
+
+    Returns:
+        Tensor: Sequence of Weighting Function.
+    """
+    upper_bound1 = abs(up[0]) * abs(reg_scale)
+    upper_bound2 = abs(up[0]) * abs(reg_scale) * 2
+    step = (upper_bound1 + 1) ** (2 / (reg_max - 2))
+    left_values = [-((step) ** i) + 1 for i in range(reg_max // 2 - 1, 0, -1)]
+    right_values = [(step) ** i - 1 for i in range(1, reg_max // 2)]
+    return torch.cat(
+        [
+            -upper_bound2,
+            torch.cat(left_values),
+            torch.zeros_like(up[0][None]),
+            torch.cat(right_values),
+            upper_bound2,
+        ],
+        0,
+    )
+
+
+def distance2bbox(points: Tensor, distance: Tensor, reg_scale: Tensor) -> Tensor:
+    """Decodes edge-distances into bounding box coordinates.
+
+    Args:
+        points (Tensor): (B, N, 4) or (N, 4) format, representing [x, y, w, h],
+                        where (x, y) is the center and (w, h) are width and height.
+        distance (Tensor): (B, N, 4) or (N, 4), representing distances from the
+                        point to the left, top, right, and bottom boundaries.
+
+        reg_scale (float): Controls the curvature of the Weighting Function.
+
+    Returns:
+        Tensor: Bounding boxes in (N, 4) or (B, N, 4) format [cx, cy, w, h].
+    """
+    reg_scale = abs(reg_scale)
+    x1 = points[..., 0] - (0.5 * reg_scale + distance[..., 0]) * (points[..., 2] / reg_scale)
+    y1 = points[..., 1] - (0.5 * reg_scale + distance[..., 1]) * (points[..., 3] / reg_scale)
+    x2 = points[..., 0] + (0.5 * reg_scale + distance[..., 2]) * (points[..., 2] / reg_scale)
+    y2 = points[..., 1] + (0.5 * reg_scale + distance[..., 3]) * (points[..., 3] / reg_scale)
+
+    bboxes = torch.stack([x1, y1, x2, y2], -1)
+    return box_convert(bboxes, in_fmt="xyxy", out_fmt="cxcywh")
+
+
+def deformable_attention_core_func_v2(
```

Comment:
Same comment about the location, maybe: otx/src/otx/algo/modules/transformer.py?